src/ordered.jl

```diff
@@ -8,35 +8,86 @@ When applying the algorithm to an image, the threshold matrix is repeatedly tile
 """
-struct OrderedDither{T<:AbstractMatrix} <: AbstractDither
+struct OrderedDither{T<:AbstractMatrix{<:Rational},R<:Real} <: AbstractDither
     mat::T
+    color_error_multiplier::R
 end
-function OrderedDither(mat; invert_map=false)
+function OrderedDither(mat; invert_map=false, color_error_multiplier=0.5)
     if invert_map
-        return OrderedDither(1 .- mat)
+        return OrderedDither(1 .- mat, color_error_multiplier)
     end
-    return OrderedDither(mat)
+    return OrderedDither(mat, color_error_multiplier)
 end
 
 function binarydither!(alg::OrderedDither, out::GenericGrayImage, img::GenericGrayImage)
     # eagerly promote to the same eltype to make for-loop faster
     FT = floattype(eltype(img))
     mat = FT.(alg.mat)
 
+    T = eltype(out)
+    black, white = T(0), T(1)
+
+    # Precompute lookup tables for modulo indexing of threshold matrix
     matsize = size(mat)
     rlookup = [mod1(i, matsize[1]) for i in 1:size(img)[1]]
     clookup = [mod1(i, matsize[2]) for i in 1:size(img)[2]]
 
-    T = eltype(out)
-    black, white = T(0), T(1)
-
     @inbounds @simd for i in CartesianIndices(img)
         r, c = Tuple(i)
         out[i] = ifelse(img[i] > mat[rlookup[r], clookup[c]], white, black)
     end
     return out
 end
 
+# Using Pattern Dithering by Thomas Knoll / Adobe Inc. Patent expired in 2019.
+# https://patents.google.com/patent/US6606166B1/en
+# Implemented according to Joel Yliluoma's pseudocode
+# https://bisqwit.iki.fi/story/howto/dither/jy/
+function colordither(
+    alg::OrderedDither,
+    img::GenericImage,
+    cs::AbstractVector{<:Pixel},
+    metric::DifferenceMetric,
+)
+    cs_lab = Lab.(cs)
+    cs_xyz = XYZ.(cs)
+
+    # Precompute lookup tables for modulo indexing of threshold matrix
+    mat = numerator.(alg.mat)
+    nmax = maximum(mat)
+    matsize = size(mat)
+    rlookup = [mod1(i, matsize[1]) for i in 1:size(img)[1]]
+    clookup = [mod1(i, matsize[2]) for i in 1:size(img)[2]]
+
+    # Allocate matrices
+    candidates = Array{UInt8}(undef, nmax)
+    index = Matrix{UInt8}(undef, size(img)...)
+
+    @inbounds for I in CartesianIndices(img)
+        r, c = Tuple(I)
+        err = zero(XYZ)
+        px = XYZ(img[I])
+
+        for j in 1:nmax
+            col = px + alg.color_error_multiplier * err
+            idx = _closest_color_idx(col, cs_lab, metric)
+            candidates[j] = idx
+            err = px - cs_xyz[idx]
+
+            # We can stop the loop if idx stayed constant:
+            if j > 2 && (candidates[j - 1] == idx)
+                candidates[(j + 1):end] .= idx
+                break
+            end
+            # TODO: also break if a series of indices repeats, e.g. [1, 2, 1, 2, ...]
+            # This would increase the performance by a lot.
+        end
+        # Sort candidates by luminance (dark to bright)
+        index[I] = sort(candidates; by=i -> cs_lab[i].l)[mat[rlookup[r], clookup[c]]]
```

Comment:
How large `nmax` could be here? If it is too large then we can maybe try `nsmallest` and/or `nextreme` from DataStructures https://juliacollections.github.io/DataStructures.jl/latest/heaps/#Functions-using-heaps-1

This is a known algorithm "select kth smallest element" and is expected to have `O(n+klogn)` complexity (I'm not an algorithm expert, I googled it.)

Reply:
Ah okay, it seems that `nmax=36` is quite small so that there's no need to apply fancy algorithms. But in place partial sort could still be useful.

```julia
julia> @btime partialsort!($x, 20) setup=(x=rand(1:100, 36));
  27.591 ns (0 allocations: 0 bytes)

julia> @btime sort($x) setup=(x=rand(1:100, 36));
  113.579 ns (1 allocation: 368 bytes)

julia> @btime sort!($x) setup=(x=rand(1:100, 36));
  73.069 ns (0 allocations: 0 bytes)
```

Reply:
A reasonable size to benchmark on is `nmax=64`. In some exceptional cases, `nmax` could be really high
```julia
julia> prod(size(Bayer(level=4).mat))
1024
```
but this algorithm just isn't really well suited to such large threshold matrices.

Reply:
Another approach is to precompute sorted colorschemes:
```julia
cs_lab = sort(Lab.(cs), by=c->c.l)
cs_xyz = XYZ.(cs_lab)
# ...
index[I] = sort(candidates)[mat[rlookup[r], clookup[c]]]
```
However, these indices then need to be mapped back to the original indices of `cs`.